diff --git a/nixos/modules/system/activation/top-level.nix b/nixos/modules/system/activation/top-level.nix
index 9e6ca75b..13ea5325 100644
--- a/nixos/modules/system/activation/top-level.nix
+++ b/nixos/modules/system/activation/top-level.nix
@@ -191,7 +191,13 @@ in
               specialisations will be ignored.
             '';
             visible = "shallow";
-            inherit (extend { modules = [ ./no-clone.nix ]; }) type;
+            inherit (extend { modules = [ ./no-clone.nix ]; specialArgs = local.config.specialArgs; }) type;
+          };
+
+          options.specialArgs = mkOption {
+            type = types.attrsOf types.anything;
+            default = { };
+            description = "Additional <literal>specialArgs</literal> to pass to modules evaluated for this specialisation.";
           };
         })
       );
